drafts/2019-10-01-rewriting-irewritable.md

```diff
@@ -0,0 +1,448 @@
+
+So `Span` thoroughly solves the safety issue with the array-oriented API. I don't have to worry about mutating a `Span` which got stored inside a client object, because the `Span` can't be stored! This allows Sawmill to safely reuse `Span`s, rather than allocating a new array for each `GetChildren`/`SetChildren` call.
+
+So Sawmill's methods like `RewriteChildren` can be implemented without allocating memory. In this example I'm using [`ArrayPool`](https://docs.microsoft.com/en-us/dotnet/api/system.buffers.arraypool-1?view=netcore-3.0) to avoid creating a new array for every `RewriteChildren` call.
+
+```csharp
+public static T RewriteChildren<T>(this T value, Func<T, T> transformer) where T : IRewritable<T>
+{
+    var count = value.CountChildren();
+    var array = ArrayPool<T>.Shared.Rent(count);
+    // ArrayPool can return arrays bigger than you asked for
+    var span = array.AsSpan().Slice(count);
+
+    value.GetChildren(span);
+    
+    for (var i = 0; i < span.Length; i++)
+    {
+        span[i] = transformer(span[i]);
+    }
+
+    var result = value.SetChildren(newChildren);
+    ArrayPool<T>.Shared.Return(array);
+    return result;
+}
+```
+
+The main reason for having `RewriteChildren` be a method on the interface was to avoid allocating memory (for the `GetChildren` calls). So we don't need it on the interface any more --- this extension method serves as a single universal implementation. Likewise, `Children<T>`'s purpose was also to avoid allocating memory, so we can do away with it too.
+
+
+## Relieving `ArrayPool` pressure
+
+There's one operational problem with this implementation: it can end up renting a large number of arrays from the pool. `Rewrite`, which applies a transformation function to every node in a tree (not just one layer), is implemented something like this:
+
+```csharp
+public static T Rewrite(this T value, Func<T, T> transformer) where T : IRewritable<T>
+    => transformer(t.RewriteChildren(child => child.Rewrite(transformer)));
+```
+
+The lambda which is passed to `RewriteChildren` contains a recursive call to `Rewrite`. Let's think through the operational behaviour of `Rewrite`:
+
+1. `Rewrite` calls `RewriteChildren`
+2. `RewriteChildren` rents an array from the array pool
+3. `RewriteChildren` calls the `child => child.Rewrite(transformer)` lambda function for each child
+4. The lambda function recursively calls `Rewrite`; steps 1-3 are repeated until you encounter a node with no children
+5. `RewriteChildren` calls `SetChildren` and returns its array to the array pool
+6. `RewriteChildren` returns and step 5 is repeated as you return up the call stack
+
+Since steps 1-3 are repeated before step 5 happens, you can end up renting many arrays (a number equal to the height of the tree) before returning any of them to the pool. So the array pool could run out of arrays!
```

Comment:
`ArrayPool<T>` won't ever run out of arrays, in empty cases it just allocates a brand new one.